File: src/main/java/com/example/demo/service/TimesheetService.java
package com.example.demo.service;

import com.example.demo.mapper.TimesheetEntryMapper;
import com.example.demo.model.TimesheetEntry;
import com.example.demo.model.TimesheetSaveCommand;
import com.example.demo.util.DbUtils;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.time.LocalTime;
import java.util.ArrayList;
import java.util.List;

@Slf4j
@Service
@RequiredArgsConstructor
public class TimesheetService {

    private final TimesheetEntryMapper timesheetEntryMapper;
    private final ApplicationEventPublisher eventPublisher;


    private TimesheetEntry applyCalc(TimesheetEntry entry) {
        TimesheetEval eval = TimesheetEvaluator.evaluate(
                entry.getStartTime(),
                entry.getEndTime(),
                entry.getBreakMinutes()
        );
        if (!eval.isValid()) {
            log.warn("Timesheet validation warnings for entry {}: {}",
                    entry.getId(),
                    String.join(";", eval.errors()));
        }
        TimesheetEntry updated = copyEntry(entry);
        updated.setDurationMinutes(eval.durationMinutes());
        updated.setWorkingMinutes(eval.workingMinutes());
        return updated;
    }

    @Transactional
    public TimesheetEntry clockIn(String userName, LocalTime now) {
        LocalDate today = LocalDate.now();
        TimesheetEntry existing = DbUtils.dbCall(
                "selectByUserAndDate",
                () -> timesheetEntryMapper.selectByUserAndDate(userName, today),
                userName, today
        );
        if (existing != null) {
            if (existing.getStartTime() == null) {
                TimesheetEntry updated = copyEntry(existing);
                updated.setStartTime(now);
                TimesheetEntry finalUpdated = applyCalc(updated);
                int updatedCount = DbUtils.dbCall(
                        "updateTimes",
                        () -> timesheetEntryMapper.updateTimes(finalUpdated),
                        userName, today, existing.getId()
                );
                if (updatedCount == 0) {
                    throw new TimesheetConflictException("同時更新により保存できませんでした");
                }
                return finalUpdated;
            }
            if (existing.getEndTime() == null) {
                throw new TimesheetConflictException("既に勤務中です: clock-out が必要");
            }
            return applyCalc(existing);
        }
        TimesheetEntry entry = new TimesheetEntry();
        entry.setWorkDate(today);
        entry.setUserName(userName);
        entry.setStartTime(now);
        TimesheetEntry finalEntry = applyCalc(entry);
        try {
            DbUtils.dbCall("insert", () -> timesheetEntryMapper.insert(finalEntry), userName, today);
        } catch (Exception ex) {
            TimesheetEntry nowExisting = DbUtils.dbCall(
                    "selectByUserAndDate (insert-catch)",
                    () -> timesheetEntryMapper.selectByUserAndDate(userName, today),
                    userName, today
            );
            if (nowExisting == null) {
                throw ex;
            }
            if (nowExisting.getStartTime() == null) {
                TimesheetEntry updated2 = copyEntry(nowExisting);
                updated2.setStartTime(now);
                TimesheetEntry finalUpdated2 = applyCalc(updated2);
                int updatedCount = DbUtils.dbCall(
                        "updateTimes (insert-catch)",
                        () -> timesheetEntryMapper.updateTimes(finalUpdated2),
                        userName, today, nowExisting.getId()
                );
                if (updatedCount == 0) {
                    throw new TimesheetConflictException("同時更新により保存できませんでした");
                }
                return finalUpdated2;
            }
            if (nowExisting.getEndTime() == null) {
                throw new TimesheetConflictException("既に勤務中です: clock-out が必要");
            }
            return applyCalc(nowExisting);
        }
        eventPublisher.publishEvent(new TimesheetUpdatedEvent(userName, today));
        return finalEntry;
    }

    @Transactional
    public TimesheetEntry clockOut(String userName, LocalTime now) {
        LocalDate today = LocalDate.now();
        TimesheetEntry existing = DbUtils.dbCall(
                "selectByUserAndDate",
                () -> timesheetEntryMapper.selectByUserAndDate(userName, today),
                userName, today
        );
        if (existing == null) {
            throw new TimesheetNotFoundException("本日のタイムシートがありません");
        }

        if (existing.getEndTime() != null) {
            return applyCalc(existing);
        }
        if (existing.getStartTime() == null) {
            throw new TimesheetConflictException("clock-in が未実施です");
        }

        TimesheetEntry updated = copyEntry(existing);
        updated.setEndTime(now);
        TimesheetEntry finalUpdated = applyCalc(updated);
        int updatedCount = DbUtils.dbCall(
                "updateTimes",
                () -> timesheetEntryMapper.updateTimes(finalUpdated),
                userName, today, existing.getId()
        );
        if (updatedCount == 0) {
            throw new TimesheetConflictException("同時更新により保存できませんでした");
        }
        eventPublisher.publishEvent(new TimesheetUpdatedEvent(userName, today));
        return finalUpdated;
    }

    public TimesheetEntry getToday(String userName) {
        LocalDate today = LocalDate.now();
        TimesheetEntry entry = DbUtils.dbCall(
                "selectByUserAndDate",
                () -> timesheetEntryMapper.selectByUserAndDate(userName, today),
                userName, today
        );
        return entry != null ? applyCalc(entry) : null;
    }

    public List<TimesheetEntry> list(String userName, LocalDate from, LocalDate to) {
        if (from.isAfter(to)) {
            throw new IllegalArgumentException("from は to より後ろにできません");
        }
        List<TimesheetEntry> entries = DbUtils.dbCall(
                "selectByUserAndRange",
                () -> timesheetEntryMapper.selectByUserAndRange(userName, from, to),
                userName, from, to
        );
        List<TimesheetEntry> calculatedEntries = new ArrayList<>(entries.size());
        int invalidCount = 0;
        for (TimesheetEntry e : entries) {
            TimesheetEval eval = TimesheetEvaluator.evaluate(e.getStartTime(), e.getEndTime(), e.getBreakMinutes());
            if (!eval.isValid()) {
                invalidCount++;
                log.debug("Entry {} validation issues: {}", e.getId(), String.join(";", eval.errors()));
            }
            TimesheetEntry updated = copyEntry(e);
            updated.setDurationMinutes(eval.durationMinutes());
            updated.setWorkingMinutes(eval.workingMinutes());
            calculatedEntries.add(updated);
        }
        log.info(
                "Timesheet entries user={} range={}..{} total={} invalid={}",
                userName, from, to, entries.size(), invalidCount
        );
        return calculatedEntries;
    }

    @Transactional
    public TimesheetEntry updateNote(String userName, String note) {
        LocalDate today = LocalDate.now();
        TimesheetEntry existing = DbUtils.dbCall(
                "selectByUserAndDate",
                () -> timesheetEntryMapper.selectByUserAndDate(userName, today),
                userName, today
        );
        if (existing == null) {
            throw new TimesheetNotFoundException("本日のタイムシートがありません");
        }
        DbUtils.dbCall(
                "updateNote",
                () -> timesheetEntryMapper.updateNote(existing.getId(), note),
                existing.getId(), userName, today
        );
        eventPublisher.publishEvent(new TimesheetUpdatedEvent(userName, today));
        TimesheetEntry updated = copyEntry(existing);
        updated.setNote(note);
        return applyCalc(updated);
    }

    @Transactional
    public TimesheetEntry addNoteToEntry(String userName, String note) {
        LocalDate today = LocalDate.now();
        TimesheetEntry existing = DbUtils.dbCall(
                "selectByUserAndDate",
                () -> timesheetEntryMapper.selectByUserAndDate(userName, today),
                userName, today
        );
        if (existing == null) {
            throw new TimesheetNotFoundException("No timesheet entry found for today");
        }
        DbUtils.dbCall(
                "updateNote",
                () -> timesheetEntryMapper.updateNote(existing.getId(), note),
                existing.getId(), userName, today
        );
        eventPublisher.publishEvent(new TimesheetUpdatedEvent(userName, today));
        TimesheetEntry updated = copyEntry(existing);
        updated.setNote(note);
        return applyCalc(updated);
    }

    @Transactional
    public TimesheetEntry saveOrUpdate(
            String userName,
            LocalDate workDate,
            LocalTime startTime,
            LocalTime endTime,
            Integer breakMinutes,
            boolean force,
            boolean holidayWork,
            String workLocation
    ) {
        TimesheetEntry existing = DbUtils.dbCall(
                "selectByUserAndDate",
                () -> timesheetEntryMapper.selectByUserAndDate(userName, workDate),
                userName, workDate
        );
        if (existing != null) {
            TimesheetEntry merged = copyEntry(existing);
            merged.setStartTime(startTime != null ? startTime : existing.getStartTime());
            merged.setEndTime(endTime != null ? endTime : existing.getEndTime());
            merged.setBreakMinutes(breakMinutes != null ? breakMinutes : existing.getBreakMinutes());
            merged.setHolidayWork(holidayWork);
            merged.setWorkLocation(workLocation != null ? workLocation : existing.getWorkLocation());

            TimesheetEntry recalced = applyCalc(merged);
            int updatedCount = DbUtils.dbCall(
                    "updateTimes/updateTimesForce",
                    () -> force ? timesheetEntryMapper.updateTimesForce(recalced)
                            : timesheetEntryMapper.updateTimes(recalced),
                    recalced.getId(), userName, workDate
            );
            if (updatedCount == 0) {
                throw new TimesheetConflictException("同時更新により保存できませんでした");
            }
            eventPublisher.publishEvent(new TimesheetUpdatedEvent(userName, workDate));
            return recalced;
        }

        TimesheetEntry createdBase = new TimesheetEntry();
        createdBase.setWorkDate(workDate);
        createdBase.setUserName(userName);
        createdBase.setStartTime(startTime);
        createdBase.setEndTime(endTime);
        createdBase.setBreakMinutes(breakMinutes);
        createdBase.setHolidayWork(holidayWork);
        createdBase.setWorkLocation(workLocation);
        TimesheetEntry created = applyCalc(createdBase);

        try {
            DbUtils.dbCall("insert", () -> timesheetEntryMapper.insert(created), userName, workDate);
        } catch (Exception ex) {
            TimesheetEntry nowExisting = DbUtils.dbCall(
                    "selectByUserAndDate (insert-catch)",
                    () -> timesheetEntryMapper.selectByUserAndDate(userName, workDate),
                    userName, workDate
            );
            if (nowExisting == null) {
                throw ex;
            }
            TimesheetEntry merged = copyEntry(nowExisting);
            merged.setStartTime(startTime != null ? startTime : nowExisting.getStartTime());
            merged.setEndTime(endTime != null ? endTime : nowExisting.getEndTime());
            merged.setBreakMinutes(breakMinutes != null ? breakMinutes : nowExisting.getBreakMinutes());
            merged.setHolidayWork(holidayWork);
            merged.setWorkLocation(workLocation != null ? workLocation : nowExisting.getWorkLocation());

            TimesheetEntry recalced = applyCalc(merged);
            int updatedCount = DbUtils.dbCall(
                    "updateTimes/updateTimesForce (insert-catch)",
                    () -> force ? timesheetEntryMapper.updateTimesForce(recalced)
                            : timesheetEntryMapper.updateTimes(recalced),
                    recalced.getId(), userName, workDate
            );
            if (updatedCount == 0) {
                throw new TimesheetConflictException("同時更新により保存できませんでした");
            }
            eventPublisher.publishEvent(new TimesheetUpdatedEvent(userName, workDate));
            return recalced;
        }

        eventPublisher.publishEvent(new TimesheetUpdatedEvent(userName, workDate));
        return created;
    }

    @Transactional
    public TimesheetEntry saveOrUpdateWithFlags(TimesheetSaveCommand cmd) {
        TimesheetEntry existing = DbUtils.dbCall(
                "selectByUserAndDate",
                () -> timesheetEntryMapper.selectByUserAndDate(cmd.userName(), cmd.workDate()),
                cmd.userName(), cmd.workDate()
        );
        if (existing != null) {
            TimesheetEntry merged = copyEntry(existing);
            merged.setStartTime(cmd.startProvided() ? cmd.startTime() : existing.getStartTime());
            merged.setEndTime(cmd.endProvided() ? cmd.endTime() : existing.getEndTime());
            merged.setBreakMinutes(cmd.breakProvided() ? cmd.breakMinutes() : existing.getBreakMinutes());
            merged.setHolidayWork(cmd.holidayWork());
            merged.setNote(cmd.noteProvided() ? cmd.note() : existing.getNote());
            merged.setWorkLocation(cmd.clearWorkLocation() ? null : (cmd.workLocation() != null ? cmd.workLocation() : existing.getWorkLocation()));

            merged.setIrregularWorkType(cmd.clearIrregular() ? null : (cmd.irregularWorkType() != null ? cmd.irregularWorkType() : existing.getIrregularWorkType()));
            merged.setIrregularWorkDesc(cmd.clearIrregular() ? null : (cmd.irregularWorkDesc() != null ? cmd.irregularWorkDesc() : existing.getIrregularWorkDesc()));
            merged.setIrregularWorkData(cmd.clearIrregular() ? null : (cmd.irregularWorkData() != null ? cmd.irregularWorkData() : existing.getIrregularWorkData()));
            merged.setLateTime(cmd.clearLate() ? null : (cmd.lateTime() != null ? cmd.lateTime() : existing.getLateTime()));
            merged.setLateDesc(cmd.clearLate() ? null : (cmd.lateDesc() != null ? cmd.lateDesc() : existing.getLateDesc()));
            merged.setEarlyTime(cmd.clearEarly() ? null : (cmd.earlyTime() != null ? cmd.earlyTime() : existing.getEarlyTime()));
            merged.setEarlyDesc(cmd.clearEarly() ? null : (cmd.earlyDesc() != null ? cmd.earlyDesc() : existing.getEarlyDesc()));
            merged.setFreeNote(cmd.clearFreeNote() ? null : (cmd.freeNote() != null ? cmd.freeNote() : existing.getFreeNote()));
            merged.setPaidLeave(cmd.clearPaidLeave() ? null : (cmd.paidLeave() != null ? cmd.paidLeave() : existing.getPaidLeave()));

            TimesheetEntry recalced = applyCalc(merged);
            int updatedCount = DbUtils.dbCall(
                    "updateTimes/updateTimesForce",
                    () -> cmd.force() ? timesheetEntryMapper.updateTimesForce(recalced)
                            : timesheetEntryMapper.updateTimes(recalced),
                    recalced.getId(), cmd.userName(), cmd.workDate()
            );
            if (updatedCount == 0) {
                throw new TimesheetConflictException("同時更新により保存できませんでした");
            }
            eventPublisher.publishEvent(new TimesheetUpdatedEvent(cmd.userName(), cmd.workDate()));
            return recalced;
        }

        TimesheetEntry createdBase = new TimesheetEntry();
        createdBase.setWorkDate(cmd.workDate());
        createdBase.setUserName(cmd.userName());
        createdBase.setStartTime(cmd.startTime());
        createdBase.setEndTime(cmd.endTime());
        createdBase.setBreakMinutes(cmd.breakMinutes());
        createdBase.setHolidayWork(cmd.holidayWork());
        createdBase.setNote(cmd.note());
        createdBase.setWorkLocation(cmd.workLocation());
        createdBase.setIrregularWorkType(cmd.irregularWorkType());
        createdBase.setIrregularWorkDesc(cmd.irregularWorkDesc());
        createdBase.setIrregularWorkData(cmd.irregularWorkData());
        createdBase.setLateTime(cmd.lateTime());
        createdBase.setLateDesc(cmd.lateDesc());
        createdBase.setEarlyTime(cmd.earlyTime());
        createdBase.setEarlyDesc(cmd.earlyDesc());
        createdBase.setFreeNote(cmd.freeNote());
        createdBase.setPaidLeave(cmd.paidLeave());

        TimesheetEntry created = applyCalc(createdBase);
        DbUtils.dbCall("insert", () -> timesheetEntryMapper.insert(created), cmd.userName(), cmd.workDate());
        eventPublisher.publishEvent(new TimesheetUpdatedEvent(cmd.userName(), cmd.workDate()));
        return created;
    }

    private TimesheetEntry copyEntry(TimesheetEntry source) {
        return source.toBuilder().build();
    }
}

File: src/main/kotlin/com/example/demo/service/TimesheetService.kt
package com.example.demo.service

import com.example.demo.mapper.TimesheetEntryMapper
import com.example.demo.model.TimesheetEntry
import com.example.demo.util.dbCall
import org.slf4j.LoggerFactory
import org.springframework.context.ApplicationEventPublisher
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDate
import java.time.LocalTime

class TimesheetConflictException(message: String) : RuntimeException(message)
class TimesheetNotFoundException(message: String) : RuntimeException(message)

@Service
class TimesheetService(
    private val timesheetEntryMapper: TimesheetEntryMapper,
    private val eventPublisher: ApplicationEventPublisher
) {
    companion object {
        private val logger = LoggerFactory.getLogger(TimesheetService::class.java)
    }

    private fun applyCalc(entry: TimesheetEntry): TimesheetEntry {
        val eval = TimesheetEvaluator.evaluate(entry.startTime, entry.endTime, entry.breakMinutes)
        if (!eval.isValid) {
            logger.warn("Timesheet validation warnings for entry {}: {}", entry.id, eval.errors.joinToString(";"))
        }
        return entry.copy(durationMinutes = eval.durationMinutes, workingMinutes = eval.workingMinutes)
    }

    @Transactional
    fun clockIn(userName: String, now: LocalTime = LocalTime.now()): TimesheetEntry {
        val today = LocalDate.now()
        val existing =
            dbCall("selectByUserAndDate", userName, today) { timesheetEntryMapper.selectByUserAndDate(userName, today) }
        if (existing != null) {
            if (existing.startTime == null) {
                // 補完: startTime 未設定ならセット
                val updated = applyCalc(existing.copy(startTime = now))
                val updatedCount =
                    dbCall("updateTimes", userName, today, existing.id) { timesheetEntryMapper.updateTimes(updated) }
                if (updatedCount == 0) throw TimesheetConflictException("同時更新により保存できませんでした")
                return updated
            }
            if (existing.endTime == null) {
                throw TimesheetConflictException("既に勤務中です: clock-out が必要")
            }
            // 当日完了済 -> そのまま返却
            return applyCalc(existing)
        }
        val entry = applyCalc(TimesheetEntry(workDate = today, userName = userName, startTime = now))
        try {
            dbCall("insert", userName, today) { timesheetEntryMapper.insert(entry) }
        } catch (ex: Exception) {
            // 競合で同一 user+date が挿入されていた場合、既存レコードを取得して再試行
            val nowExisting = dbCall(
                "selectByUserAndDate (insert-catch)",
                userName,
                today
            ) { timesheetEntryMapper.selectByUserAndDate(userName, today) }
                ?: throw ex
            if (nowExisting.startTime == null) {
                val updated = applyCalc(nowExisting.copy(startTime = now))
                val updatedCount = dbCall(
                    "updateTimes (insert-catch)",
                    userName,
                    today,
                    nowExisting.id
                ) { timesheetEntryMapper.updateTimes(updated) }
                if (updatedCount == 0) throw TimesheetConflictException("同時更新により保存できませんでした")
                return updated
            }
            if (nowExisting.endTime == null) {
                throw TimesheetConflictException("既に勤務中です: clock-out が必要")
            }
            return applyCalc(nowExisting)
        }
        eventPublisher.publishEvent(TimesheetUpdatedEvent(userName, today))
        return entry
    }

    @Transactional
    fun clockOut(userName: String, now: LocalTime = LocalTime.now()): TimesheetEntry {
        val today = LocalDate.now()
        val existing = dbCall("selectByUserAndDate", userName, today) {
            timesheetEntryMapper.selectByUserAndDate(userName, today)
        } ?: throw TimesheetNotFoundException("本日のタイムシートがありません")

        if (existing.endTime != null) {
            // Already clocked out
            return applyCalc(existing)
        }
        if (existing.startTime == null) {
            throw TimesheetConflictException("clock-in が未実施です")
        }

        val updated = applyCalc(existing.copy(endTime = now))
        val updatedCount =
            dbCall("updateTimes", userName, today, existing.id) { timesheetEntryMapper.updateTimes(updated) }
        if (updatedCount == 0) throw TimesheetConflictException("同時更新により保存できませんでした")
        eventPublisher.publishEvent(TimesheetUpdatedEvent(userName, today))
        return updated
    }

    fun getToday(userName: String): TimesheetEntry? {
        val entry = dbCall(
            "selectByUserAndDate",
            userName,
            LocalDate.now()
        ) { timesheetEntryMapper.selectByUserAndDate(userName, LocalDate.now()) }
        return entry?.let { applyCalc(it) }
    }

    fun list(userName: String, from: LocalDate, to: LocalDate): List<TimesheetEntry> {
        require(!from.isAfter(to)) { "from は to より後ろにできません" }
        val entries = dbCall("selectByUserAndRange", userName, from, to) {
            timesheetEntryMapper.selectByUserAndRange(
                userName,
                from,
                to
            )
        }
        val calculatedEntries = ArrayList<TimesheetEntry>(entries.size)
        var invalidCount = 0
        for (e in entries) {
            val eval = TimesheetEvaluator.evaluate(e.startTime, e.endTime, e.breakMinutes)
            if (!eval.isValid) {
                invalidCount++
                logger.debug("Entry {} validation issues: {}", e.id, eval.errors.joinToString(";"))
            }
            calculatedEntries += e.copy(durationMinutes = eval.durationMinutes, workingMinutes = eval.workingMinutes)
        }
        logger.info(
            "Timesheet entries user={} range={}..{} total={} invalid={}",
            userName,
            from,
            to,
            entries.size,
            invalidCount
        )
        return calculatedEntries
    }

    @Transactional
    fun updateNote(userName: String, note: String): TimesheetEntry {
        val today = LocalDate.now()
        val existing =
            dbCall("selectByUserAndDate", userName, today) { timesheetEntryMapper.selectByUserAndDate(userName, today) }
                ?: throw TimesheetNotFoundException("本日のタイムシートがありません")
        dbCall("updateNote", existing.id, userName, today) { timesheetEntryMapper.updateNote(existing.id!!, note) }
        eventPublisher.publishEvent(TimesheetUpdatedEvent(userName, today))
        return applyCalc(existing.copy(note = note))
    }

    @Transactional
    fun addNoteToEntry(userName: String, note: String): TimesheetEntry {
        val today = LocalDate.now()
        val existing = dbCall("selectByUserAndDate", userName, today) {
            timesheetEntryMapper.selectByUserAndDate(userName, today)
        } ?: throw TimesheetNotFoundException("No timesheet entry found for today")

        dbCall("updateNote", existing.id, userName, today) {
            timesheetEntryMapper.updateNote(existing.id!!, note)
        }
        eventPublisher.publishEvent(TimesheetUpdatedEvent(userName, today))
        return applyCalc(existing.copy(note = note))
    }

    @Transactional
    fun saveOrUpdate(
        userName: String,
        workDate: LocalDate,
        startTime: LocalTime?,
        endTime: LocalTime?,
        breakMinutes: Int? = null,
        force: Boolean = false,
        holidayWork: Boolean = false,
        workLocation: String? = null
    ): TimesheetEntry {
        val existing = dbCall("selectByUserAndDate", userName, workDate) {
            timesheetEntryMapper.selectByUserAndDate(
                userName,
                workDate
            )
        }
        return if (existing != null) {
            val merged = existing.copy(
                startTime = startTime ?: existing.startTime,
                endTime = endTime ?: existing.endTime,
                breakMinutes = breakMinutes ?: existing.breakMinutes,
                holidayWork = holidayWork,
                workLocation = workLocation ?: existing.workLocation
            )
            val recalced = applyCalc(merged)
            val updatedCount = dbCall("updateTimes/updateTimesForce", recalced.id, userName, workDate) {
                if (force) timesheetEntryMapper.updateTimesForce(recalced) else timesheetEntryMapper.updateTimes(
                    recalced
                )
            }
            if (updatedCount == 0) throw TimesheetConflictException("同時更新により保存できませんでした")
            eventPublisher.publishEvent(TimesheetUpdatedEvent(userName, workDate))
            recalced
        } else {
            val createdBase = TimesheetEntry(
                workDate = workDate,
                userName = userName,
                startTime = startTime,
                endTime = endTime,
                breakMinutes = breakMinutes,
                holidayWork = holidayWork,
                workLocation = workLocation
            )
            val created = applyCalc(createdBase)
            try {
                dbCall("insert", userName, workDate) { timesheetEntryMapper.insert(created) }
            } catch (ex: Exception) {
                // 競合で既にレコードが挿入されていた場合は、再取得して update を試みる
                val nowExisting = dbCall(
                    "selectByUserAndDate (insert-catch)",
                    userName,
                    workDate
                ) { timesheetEntryMapper.selectByUserAndDate(userName, workDate) }
                    ?: throw ex
                val merged = nowExisting.copy(
                    startTime = startTime ?: nowExisting.startTime,
                    endTime = endTime ?: nowExisting.endTime,
                    breakMinutes = breakMinutes ?: nowExisting.breakMinutes,
                    holidayWork = holidayWork,
                    workLocation = workLocation ?: nowExisting.workLocation
                )
                val recalced = applyCalc(merged)
                val updatedCount =
                    dbCall("updateTimes/updateTimesForce (insert-catch)", recalced.id, userName, workDate) {
                        if (force) timesheetEntryMapper.updateTimesForce(recalced) else timesheetEntryMapper.updateTimes(
                            recalced
                        )
                    }
                if (updatedCount == 0) throw TimesheetConflictException("同時更新により保存できませんでした")
                eventPublisher.publishEvent(TimesheetUpdatedEvent(userName, workDate))
                return recalced
            }
            eventPublisher.publishEvent(TimesheetUpdatedEvent(userName, workDate))
            created
        }
    }

    /**
     * Variant of saveOrUpdate that accepts explicit "provided" flags for start/end/break.
     * When a provided flag is true, the corresponding value (which may be null) will be used
     * (allowing callers to explicitly clear a field by providing null). When the provided flag
     * is false, the existing DB value is preserved.
     */
    @Transactional
    fun saveOrUpdateWithFlags(
        userName: String,
        workDate: LocalDate,
        startProvided: Boolean,
        startTime: LocalTime?,
        endProvided: Boolean,
        endTime: LocalTime?,
        breakProvided: Boolean,
        breakMinutes: Int?,
        force: Boolean = false,
        holidayWork: Boolean = false,
        workLocation: String? = null
    ): TimesheetEntry {
        val existing = dbCall("selectByUserAndDate", userName, workDate) {
            timesheetEntryMapper.selectByUserAndDate(
                userName,
                workDate
            )
        }
        return if (existing != null) {
            val newStart = if (startProvided) startTime else existing.startTime
            val newEnd = if (endProvided) endTime else existing.endTime
            val newBreak = if (breakProvided) breakMinutes else existing.breakMinutes
            val newHoliday = holidayWork
            val newWorkLocation = workLocation ?: existing.workLocation

            // If nothing changed, avoid DB write and just return evaluated existing
            val nothingChanged =
                (existing.startTime == newStart) && (existing.endTime == newEnd) && (existing.breakMinutes == newBreak) && (existing.holidayWork == newHoliday) && (existing.workLocation == newWorkLocation)
            if (nothingChanged && !force) {
                return applyCalc(existing)
            }

            val merged = existing.copy(
                startTime = newStart,
                endTime = newEnd,
                breakMinutes = newBreak,
                holidayWork = newHoliday,
                workLocation = newWorkLocation
            )
            val recalced = applyCalc(merged)
            val updatedCount = dbCall("updateTimes/updateTimesForce", recalced.id, userName, workDate) {
                if (force) timesheetEntryMapper.updateTimesForce(recalced) else timesheetEntryMapper.updateTimes(
                    recalced
                )
            }
            if (updatedCount == 0) throw TimesheetConflictException("同時更新により保存できませんでした")
            eventPublisher.publishEvent(TimesheetUpdatedEvent(userName, workDate))
            recalced
        } else {
            val createdBase = TimesheetEntry(
                workDate = workDate,
                userName = userName,
                startTime = if (startProvided) startTime else null,
                endTime = if (endProvided) endTime else null,
                breakMinutes = if (breakProvided) breakMinutes else null,
                holidayWork = holidayWork,
                workLocation = workLocation
            )
            val created = applyCalc(createdBase)
            try {
                dbCall("insert", userName, workDate) { timesheetEntryMapper.insert(created) }
            } catch (ex: Exception) {
                val nowExisting = dbCall(
                    "selectByUserAndDate (insert-catch)",
                    userName,
                    workDate
                ) { timesheetEntryMapper.selectByUserAndDate(userName, workDate) }
                    ?: throw ex
                val merged = nowExisting.copy(
                    startTime = if (startProvided) startTime else nowExisting.startTime,
                    endTime = if (endProvided) endTime else nowExisting.endTime,
                    breakMinutes = if (breakProvided) breakMinutes else nowExisting.breakMinutes,
                    holidayWork = holidayWork,
                    workLocation = workLocation ?: nowExisting.workLocation
                )
                val recalced = applyCalc(merged)
                val updatedCount =
                    dbCall("updateTimes/updateTimesForce (insert-catch)", recalced.id, userName, workDate) {
                        if (force) timesheetEntryMapper.updateTimesForce(recalced) else timesheetEntryMapper.updateTimes(
                            recalced
                        )
                    }
                if (updatedCount == 0) throw TimesheetConflictException("同時更新により保存できませんでした")
                eventPublisher.publishEvent(TimesheetUpdatedEvent(userName, workDate))
                return recalced
            }
            eventPublisher.publishEvent(TimesheetUpdatedEvent(userName, workDate))
            created
        }
    }

    /**
     * Variant of saveOrUpdate that accepts explicit "provided" flags for start/end/break.
     * When a provided flag is true, the corresponding value (which may be null) will be used
     * (allowing callers to explicitly clear a field by providing null). When the provided flag
     * is false, the existing DB value is preserved.
     *
     * For extended fields (irregularWorkType, lateTime, etc.):
     * - Pass the actual value to update
     * - Pass empty string "" to clear the field
     * - Pass null to keep existing value
     */
    @Transactional
    fun saveOrUpdateWithFlags(
        userName: String,
        workDate: LocalDate,
        startProvided: Boolean,
        startTime: LocalTime?,
        endProvided: Boolean,
        endTime: LocalTime?,
        breakProvided: Boolean,
        breakMinutes: Int?,
        force: Boolean = false,
        holidayWork: Boolean = false,
        noteProvided: Boolean,
        note: String?,
        workLocation: String? = null,
        irregularWorkType: String? = null,
        irregularWorkDesc: String? = null,
        irregularWorkData: String? = null,
        lateTime: String? = null,
        lateDesc: String? = null,
        earlyTime: String? = null,
        earlyDesc: String? = null,
        paidLeave: String? = null,
        // フラグ: 明示的にクリアするかどうか
        clearIrregular: Boolean = false,
        clearLate: Boolean = false,
        clearEarly: Boolean = false
    ): TimesheetEntry {
        val existing = dbCall("selectByUserAndDate", userName, workDate) {
            timesheetEntryMapper.selectByUserAndDate(userName, workDate)
        }
        return if (existing != null) {
            val merged = existing.copy(
                startTime = if (startProvided) startTime else existing.startTime,
                endTime = if (endProvided) endTime else existing.endTime,
                breakMinutes = if (breakProvided) breakMinutes else existing.breakMinutes,
                holidayWork = holidayWork,
                note = if (noteProvided) note else existing.note,
                workLocation = workLocation ?: existing.workLocation,
                // 拡張フィールド: clearフラグがtrueならnull、そうでなければ値があれば更新、なければ既存値
                irregularWorkType = if (clearIrregular) null else (irregularWorkType ?: existing.irregularWorkType),
                irregularWorkDesc = if (clearIrregular) null else (irregularWorkDesc ?: existing.irregularWorkDesc),
                irregularWorkData = if (clearIrregular) null else (irregularWorkData ?: existing.irregularWorkData),
                lateTime = if (clearLate) null else (lateTime ?: existing.lateTime),
                lateDesc = if (clearLate) null else (lateDesc ?: existing.lateDesc),
                earlyTime = if (clearEarly) null else (earlyTime ?: existing.earlyTime),
                earlyDesc = if (clearEarly) null else (earlyDesc ?: existing.earlyDesc),
                paidLeave = paidLeave ?: existing.paidLeave
            )
            val recalced = applyCalc(merged)
            val updatedCount = dbCall("updateTimes/updateTimesForce", recalced.id, userName, workDate) {
                if (force) timesheetEntryMapper.updateTimesForce(recalced) else timesheetEntryMapper.updateTimes(
                    recalced
                )
            }
            if (updatedCount == 0) throw TimesheetConflictException("同時更新により保存できませんでした")
            eventPublisher.publishEvent(TimesheetUpdatedEvent(userName, workDate))
            recalced
        } else {
            val createdBase = TimesheetEntry(
                workDate = workDate,
                userName = userName,
                startTime = startTime,
                endTime = endTime,
                breakMinutes = breakMinutes,
                holidayWork = holidayWork,
                note = note,
                workLocation = workLocation,
                irregularWorkType = irregularWorkType,
                irregularWorkDesc = irregularWorkDesc,
                irregularWorkData = irregularWorkData,
                lateTime = lateTime,
                lateDesc = lateDesc,
                earlyTime = earlyTime,
                earlyDesc = earlyDesc,
                paidLeave = paidLeave
            )
            val created = applyCalc(createdBase)
            dbCall("insert", userName, workDate) { timesheetEntryMapper.insert(created) }
            eventPublisher.publishEvent(TimesheetUpdatedEvent(userName, workDate))
            created
        }
    }

}
